File: rust/src/heap_sort.rs
// Code originally forked from https://github.com/diptangsu/Sorting-Algorithms/
include!("constants.rs");
use std::time::Instant;

fn move_down<T: Ord>(array: &mut [T], mut root: usize) {
    let last = array.len() - 1;
    loop {
        let left = 2 * root + 1;
        if left > last {
            break;
        }
        let right = left + 1;
        let child = if right <= last && array[right] > array[left] {
            right
        } else {
            left
        };
        if array[child] > array[root] {
            array.swap(root, child);
        }
        root = child;
    }
}

pub fn heap_sort<T: Ord>(array: &mut [T]) {
    if array.len() <= 1 {
        return;
    }
    let parent = (array.len() - 2) / 2;
    for i in (0..=parent).rev() {
        move_down(array, i);
    }
    for end in (1..array.len()).rev() {
        array.swap(0, end);
        move_down(&mut array[..end], 0);
    }
}

fn main() {
    unsafe {
        for _x in 0..50 {
            let start = Instant::now();
            heap_sort(&mut SET1);
            let elapsed = start.elapsed();
            println!("{:?}", elapsed);
        }
    }
}

File: python/heap_sort.py
# Code originally forked from https://github.com/diptangsu/Sorting-Algorithms/

import time
from constants import *

def move_down(array, n, root):
    largest = root
    left = 2 * root + 1
    right = 2 * root + 2

    if left < n and array[root] < array[left]:
        largest = left

    if right < n and array[largest] < array[right]:
        largest = right

    if largest != root:
        array[root], array[largest] = array[largest], array[root]
        move_down(array, n, largest)

def heap_sort(array):
    n = len(array)

    for i in range(n, -1, -1):
        move_down(array, n, i)

    for i in range(n - 1, 0, -1):
        array[i], array[0] = array[0], array[i]
        move_down(array, i, 0)

if __name__ == '__main__':
    for i in range(50):
        array = set1()
        start = time.time() * 1000
        heap_sort(array)
        end = time.time() * 1000
        elapsed = end - start
        print(elapsed)
